### DISCUSSION TITLE
Cannot start a runtime within a runtime with #[tokio::test]

### DISCUSSION THREAD

**Discussion by cryptoquick** [tokio-rs/tokio#4779] [General]
I'm trying to make a test that makes a couple of async calls within a library I'm building to things that use reqwest and stuff under the hood. I'm getting the error, `Cannot start a runtime from within a runtime`. What's unfortunate, though, is that I won't benefit from `?` error bubbling in my test if I just try to spawn the thread there, like this:

```rust
#[tokio::test]
async fn my_test() -> Result<()> {
    do_a_thing()?;

    let result = do_some_async_thing().await?;

    assert_eq!(result, "success");

    Ok(())
}
```

I tried anyway, and then I got issues `generator cannot be sent between threads safely`:

```rust
#[test]
fn my_test() -> Result<()> {
    do_a_thing()?;

    thread::spawn(|| async {
        let result = do_some_async_thing()
        .await
        .unwrap();

        assert_eq!(result, "success");
    })
    .join()
    .unwrap();

    Ok(())
}
```

What's the best way to solve this issue?

**Comment by Noah-Kennedy** [comment]
I'm assuming that your library spawns a tokio runtime internally.

You shouldn't do this.

It's generally far better to just make functions async than to try and supply an executor like this, because you can't guarantee that you aren't being called from within the context of another executor.

**Reply by cryptoquick** [reply]
Actually, I only added tokio for testing as a dev-dependency, I wasn't using it internally, but some downstream libraries might be. Is there a way to tell what library it might be?

**Reply by Noah-Kennedy** [reply]
Can you exclude the full panic trace?
